bootstrap/pkg/kfapp/gcp/gcp.go

```diff
@@ -2025,6 +1991,46 @@ func (gcp *Gcp) setGcpPluginDefaults() error {
 		}
 	}
 
+	// Set the email
+	if gcp.kfDef.Spec.Email == "" && gcp.gcpAccountGetter != nil {
+		email, err := gcp.gcpAccountGetter()
+		if err != nil {
+			log.Errorf("cannot get gcloud account email. Error: %v", err)
+			return err
+		}
+		gcp.kfDef.Spec.Email = strings.TrimSpace(email)
+		pluginSpec.Email = strings.TrimSpace(email)
+		log.Infof("Setting Email to default: %v", gcp.kfDef.Spec.Email)
+	} else {
+		log.Warnf("gcpAccountGetter not set; can't get default email")
+	}
+	// Set the project
+	if gcp.kfDef.Spec.Project == "" && gcp.gcpProjectGetter != nil {
+		project, err := gcp.gcpProjectGetter()
+		if err != nil {
+			log.Errorf("cannot get gcloud project. Error: %v", err)
+			return err
+		}
+		gcp.kfDef.Spec.Project = strings.TrimSpace(project)
+		pluginSpec.Project = strings.TrimSpace(project)
+		log.Infof("Setting Project to default: %v", gcp.kfDef.Spec.Project)
+	} else {
+		log.Warnf("gcpProjectGetter not set; can't get default project")
+	}
+	// Set the zone
+	if gcp.kfDef.Spec.Zone == "" && gcp.gcpZoneGetter != nil {
```

Comment:
For v1beta1, this check should be based on plugInSpec.Project.

Otherwise, there is a bug: any v1beta1 config should not set `kfDef.Spec.Project/Zone`; this code will always overwrite the pluginSpec.Project even if user explicitly specify one in the plugin config.

These two different ways of setting defaults should be separated.

Reply:
no, the check is based on `KfConfig`, not `KfDef`, tho the naming here is confusing.  those values are copied here:
https://github.com/kubeflow/kubeflow/blob/70d3119fc5f075ff9cf715e6bfe0a8ce00e05736/bootstrap/pkg/apis/apps/configconverters/v1beta1.go#L110